extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcProvider.java

```diff
@@ -0,0 +1,260 @@
+            LOG.debugf("Token verification has failed: %s", detail);
+            return Uni.createFrom().failure(ex);
+        }
+        return Uni.createFrom().item(new TokenVerificationResult(OidcUtils.decodeJwtContent(token), null));
+    }
+
+    public Uni<TokenVerificationResult> refreshJwksAndVerifyJwtToken(String token) {
+        return keyResolver.refresh().onItem().transformToUni(new Function<Void, Uni<? extends TokenVerificationResult>>() {
+
+            @Override
+            public Uni<? extends TokenVerificationResult> apply(Void t) {
+                return verifyJwtToken(token);
+            }
+
+        });
+    }
+
+    public Uni<TokenVerificationResult> introspectToken(String token) {
+        return client.introspectToken(token).onItemOrFailure()
+                .transform(new BiFunction<JsonObject, Throwable, TokenVerificationResult>() {
+
+                    @Override
+                    public TokenVerificationResult apply(JsonObject jsonObject, Throwable t) {
+                        if (t != null) {
+                            throw new AuthenticationFailedException(t);
+                        }
+                        if (!Boolean.TRUE.equals(jsonObject.getBoolean(OidcConstants.INTROSPECTION_TOKEN_ACTIVE))) {
+                            throw new AuthenticationFailedException();
+                        }
+                        Long exp = jsonObject.getLong(OidcConstants.INTROSPECTION_TOKEN_EXP);
+                        if (exp != null) {
+                            final int lifespanGrace = client.getOidcConfig().token.lifespanGrace.isPresent()
+                                    ? client.getOidcConfig().token.lifespanGrace.getAsInt()
+                                    : 0;
+                            if (System.currentTimeMillis() / 1000 > exp + lifespanGrace) {
+                                throw new AuthenticationFailedException();
+                            }
+                        }
+
+                        return new TokenVerificationResult(null, jsonObject);
+                    }
+
+                });
+    }
+
+    public void introspectToken(UniEmitter<? super TokenVerificationResult> emitter, String token) {
+        try {
+            emitter.complete(introspectToken(token).await().indefinitely());
+        } catch (Throwable t) {
+            emitter.fail(t);
+        }
+    }
+
+    public Uni<JsonObject> getUserInfo(RoutingContext vertxContext, TokenAuthenticationRequest request) {
+        String accessToken = vertxContext.get(OidcConstants.ACCESS_TOKEN_VALUE);
+        if (accessToken == null) {
+            accessToken = request.getToken().getToken();
+        }
+        return client.getUserInfo(accessToken);
+    }
+
+    public void getUserInfo(UniEmitter<? super JsonObject> emitter, RoutingContext vertxContext,
+            TokenAuthenticationRequest request) {
+        try {
+            emitter.complete(getUserInfo(vertxContext, request).await().indefinitely());
```

Comment:
I don't believe there is a need to block here. 
Why not returning the Uni directly?